server/src/main/java/io/crate/execution/engine/sort/Sort.java

```diff
@@ -22,48 +22,78 @@
 package io.crate.execution.engine.sort;
 
 import static io.crate.common.concurrent.CompletableFutures.supplyAsync;
-import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_ARRAY_HEADER;
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
 
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
+import java.util.function.LongConsumer;
+
+
+import org.apache.lucene.util.RamUsageEstimator;
 
 import io.crate.common.collections.Iterables;
 import io.crate.common.collections.Lists;
 
 public final class Sort {
 
+    @SuppressWarnings("unchecked")
     public static <T> CompletableFuture<List<T>> parallelSort(List<T> list,
+                                                              LongConsumer allocateBytes,
                                                               Comparator<? super T> comparator,
                                                               int minItemsPerThread,
                                                               int numAvailableThreads,
                                                               Executor executor) {
 
         int itemsPerThread = list.size() / Math.max(1, numAvailableThreads);
-        if (numAvailableThreads <= 1 || itemsPerThread < minItemsPerThread) {
-            list.sort(comparator);
-            return completedFuture(list);
-        }
+//        if (numAvailableThreads <= 1 || itemsPerThread < minItemsPerThread) {
+//            list.sort(comparator);
+//            allocateBytes.accept(sortOverheadBytes(list.size()));
+//            return completedFuture(list);
+//        }
+        //89 -> sublist 96-97 -> 99 00, 01 02
         List<List<T>> partitions = Lists.partition(list, itemsPerThread);
-        ArrayList<CompletableFuture<List<T>>> futures = new ArrayList<>(partitions.size());
+        CompletableFuture<List<T>>[] futures = new CompletableFuture[partitions.size()];
+        allocateBytes.accept(RamUsageEstimator.shallowSizeOf(futures));
+        int index = 0;
```

Comment:
Reason of suppressing warnings.

`Unchecked assignment: 'java.util.concurrent.CompletableFuture[]' to 'java.util.concurrent.CompletableFuture<java.util.List<T>>[]' `

We only save an extra array allocation. when converting Future List to an array. 
So maybe keep using List<Future> and avoid unsafe code + account for an extra intermediate array?